### PR TITLE
[FLINK-3647] Change StreamSource to use Processing-Time Clock Service

### PR DESCRIPTION
**Pull request by kl0u**
This PR changes the AutomaticWatermarkContext to user a pluggable processing time clock. This
allows for better testability of the code.

This PR also contains the solution to FLINK-3646.

### CODE REVIEW

**Comment by aljoscha** [comment]
This looks very good! I just have a few minor comments:

* It's very good that you fixed typos and added comments/Javadoc to previously undocumented code. I would like to see them as a separate PR, though, and keep this PR only to the feature mentioned in the Jiras. 

* You added `TestTimeServiceProvider` in two tests. I think it makes sense to move this to the same package as `DefaultTimerService` and clearly mark it as being for testing only. On a related note, it might make sense to normalize the names, right now there is `*TimeServiceProvider` and `*TimerService`.

* You added the additional method `registerTimer(final long timestamp, final Runnable target)`. I imagine it is because the sources register their watermark timers as a `Runnable`. I think it would be better to keep having only `Triggerable` here. The watermark emitters should also work as a `Triggerable`.

**Comment by kl0u** [comment]
@aljoscha I integrated the comments.
Let me know if there are any followups.

File: flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/testutils/MockRuntimeContext.java [hunk 1]
```diff
@@ -186,16 +199,19 @@ public DistributedCache getDistributedCache() {
...
+
 	@Override
 	public ScheduledFuture<?> registerTimer(final long time, final Triggerable target) {
-		if (timer == null) {
-			timer = Executors.newSingleThreadScheduledExecutor();
-		}
+		Preconditions.checkNotNull(timerService, "The processing time timer has not been initialized.");
 		
-		final long delay = Math.max(time - System.currentTimeMillis(), 0);
+		final long delay = Math.max(time - timerService.getCurrentProcessingTime(), 0);
```
Comment (aljoscha) [inline flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/testutils/MockRuntimeContext.java]: `timerService.registerTimer()` takes an absolute timestamp not a delay now.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/assigners/WindowAssignerContext.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import org.apache.flink.streaming.runtime.tasks.StreamTask;
+
+/**
+ * A context provided to the {@link WindowAssigner} that allows it to query the
+ * current processing time. This is provided to the assigner by its containing
+ * {@link org.apache.flink.streaming.runtime.operators.windowing.WindowOperator},
+ * which, in turn, gets it from the containing
+ * {@link org.apache.flink.streaming.runtime.tasks.StreamTask}.
+ */
+public abstract class WindowAssignerContext {
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/assigners/WindowAssignerContext.java]: This should be moved to `WindowAssigner`, similarly to how `TriggerContext` is an inner class of `Trigger`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/ContinuousProcessingTimeTrigger.java [hunk 3]
```diff
@@ -27,10 +27,12 @@
 import org.apache.flink.api.common.typeutils.base.LongSerializer;
 import org.apache.flink.streaming.api.windowing.time.Time;
 import org.apache.flink.streaming.api.windowing.windows.Window;
+import org.apache.flink.streaming.runtime.tasks.TimeServiceProvider;
 
 /**
- * A {@link Trigger} that continuously fires based on a given time interval. The time is the current
- * system time.
+ * A {@link Trigger} that continuously fires based on a given time interval. The current (processing)
+ * time is provided by the {@link TimeServiceProvider}
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/ContinuousProcessingTimeTrigger.java]: I think this is an implementation detail that should not necessarily be mentioned here.

**Comment by aljoscha** [comment]
@kl0u could you please close this?

**Pull request closed without merging**